src/shell/init.go

```diff
@@ -51,25 +52,111 @@ func getExecutablePath(env environment.Environment) (string, error) {
 	// On Windows, it fails when the excutable is called in MSYS2 for example
 	// which uses unix style paths to resolve the executable's location.
 	// PowerShell knows how to resolve both, so we can swap this without any issue.
-	executable = strings.ReplaceAll(executable, "\\", "/")
-	switch env.Flags().Shell {
-	case BASH, ZSH:
-		executable = strings.ReplaceAll(executable, " ", "\\ ")
-		executable = strings.ReplaceAll(executable, "(", "\\(")
-		executable = strings.ReplaceAll(executable, ")", "\\)")
+	if runtime.GOOS == "windows" {
+		executable = strings.ReplaceAll(executable, "\\", "/")
 	}
 	return executable, nil
 }
 
-func Init(env environment.Environment) string {
-	executable, err := getExecutablePath(env)
-	if err != nil {
-		return noExe
+func quotePwshStr(str string) string {
+	return fmt.Sprintf("'%s'", strings.ReplaceAll(str, "'", "''"))
+}
+
+func quotePosixStr(str string) string {
+	if len(str) == 0 {
+		return "''"
+	}
+	needQuoting := false
+	var b strings.Builder
+	for _, r := range str {
+		normal := false
+		switch r {
+		case '!', ';', '"', '(', ')', '[', ']', '{', '}', '$', '|', '&', '>', '<', '`', ' ', '#', '~', '*', '?', '=':
+			b.WriteRune(r)
+		case '\\', '\'':
+			b.WriteByte('\\')
+			b.WriteRune(r)
+		case '\a':
+			b.WriteString(`\a`)
+		case '\b':
+			b.WriteString(`\b`)
+		case '\f':
+			b.WriteString(`\f`)
+		case '\n':
+			b.WriteString(`\n`)
+		case '\r':
+			b.WriteString(`\r`)
+		case '\t':
+			b.WriteString(`\t`)
+		case '\v':
+			b.WriteString(`\v`)
+		default:
+			b.WriteRune(r)
+			normal = true
+		}
+		if !normal {
+			needQuoting = true
+		}
+	}
+	// the quoting form $'...' is used for a string includes any special characters
+	if needQuoting {
+		return fmt.Sprintf("$'%s'", b.String())
+	}
+	return b.String()
+}
+
+func quoteFishStr(str string) string {
+	if len(str) == 0 {
+		return "''"
+	}
+	needQuoting := false
+	var b strings.Builder
+	for _, r := range str {
+		normal := false
+		switch r {
+		case ';', '"', '(', ')', '[', ']', '{', '}', '$', '|', '&', '>', '<', ' ', '#', '~', '*', '?', '=':
+			b.WriteRune(r)
+		case '\\', '\'':
+			b.WriteByte('\\')
+			b.WriteRune(r)
+		default:
+			b.WriteRune(r)
+			normal = true
+		}
+		if !normal {
+			needQuoting = true
```

Comment:
Can't this be moved inside the default case and thus not need the `normal` variable? If it reaches the default once, `needQuoting` will always be true for the remaining of the loop, we only set it to true.

Reply:
I use the `normal` variable because we don't have to quote a string that doesn't contain any special characters. In this logic, only if all runes match the `default` case can we confirm the string needs no quoting. I have another way where we can count the non-special characters and check if the quantity is equal to the total. Not so graceful, though. Do you have any ideas on how to simplify this logic?

Reply:
@JanDeDobbeleer So we just keep the original, right?